conans/client/graph/range_resolver.py

```diff
@@ -147,37 +145,65 @@ def _resolve_local(self, search_ref, version_range):
+                if remote_results:
+                    self._cached_remote_found.update({search_ref: {remote.name: remote_results}})
+                remote_results = [ref for ref in remote_results
+                                  if ref.user == search_ref.user and ref.channel == search_ref.channel]
+                resolved_version = self._resolve_version(version_range, remote_results)
+                if resolved_version and not check_all_remotes:
+                    return resolved_version, remote.name
+                elif resolved_version:
+                    results.append({"remote": remote.name,
+                                    "version": resolved_version})
+        if len(results) > 0:
+            resolved_version = self._resolve_version(version_range,
+                                                     [result.get("version") for result in results])
+            for result in results:
+                if result.get("version") == resolved_version:
+                    return result.get("version"), result.get("remote")
+        else:
+            return None, None
+
+    def _resolve_cached_reference(self, search_ref, version_range, remotes, check_all_remotes):
+        resolved_ref = None
+        resolved_remote = None
         for remote in remotes.values():
             if not remotes.selected or remote == remotes.selected:
-                result = self._remote_manager.search_recipes(remote, pattern, ignorecase=False)
-                result = [ref for ref in result
-                          if ref.user == search_ref.user and ref.channel == search_ref.channel]
-                if result:
-                    return result, remote.name
-        return None, None
-
-    def _resolve_remote(self, search_ref, version_range, remotes):
-        # We should use ignorecase=False, we want the exact case!
-        found_refs, remote_name = self._cached_remote_found.get(search_ref, (None, None))
-        if found_refs is None:
+                cached_search = self._cached_remote_found.get(search_ref)
+                if cached_search:
+                    cached_refs = cached_search.get(remote.name) or []
+                    all_refs = [resolved_ref].extend(cached_refs) if resolved_ref else cached_refs
+                    resolved_ref = self._resolve_version(version_range, all_refs)
+                    if resolved_ref in cached_refs:
+                        resolved_remote = remote.name
+                    if not check_all_remotes and resolved_ref:
+                        return resolved_ref, resolved_remote
+        return resolved_ref, resolved_remote
+
+    def _resolve_remote(self, search_ref, version_range, remotes, check_all_remotes):
+        resolved_ref, remote_name = self._resolve_cached_reference(search_ref, version_range,
+                                                                   remotes, check_all_remotes)
+        if resolved_ref is None:
```

Comment:
I am not sure the caching is correct. The ``resolved_ref`` might still be ``None``, because it was not matching the previous version range query, but the ``pkg/*@user/channel`` versions from remote were indeed retrieved and cached, and we don't want to query them again to the server.